### ISSUE TITLE
Implement remaining challenges

### ISSUE THREAD

**Issue by xenolf** [xenolf/lego#2]
Currently only stubbed out.

- [x] TLS-SNI-01
- [x] DNS-01

**Comment by MalteJ** [comment]
how would you like to implement the DNS solver? With [DNS Update](https://www.ietf.org/rfc/rfc2136.txt)?

**Comment by xenolf** [comment]
Hey,

I'm truly sorry it took me so long to respond.
To this date the challenges are not really planned through implementation wise and I'm open for input on them. 

For DNS I was leaning into the direction of just asking the user to provide the DNS records as we should make no assumptions of RFC2136 support on the provider which hosts DNS for the client.

**Comment by jimmidyson** [comment]
Would really like to see DNS challenge implemented. Trying to integrate this into Kubernetes secrets & that seems like the best fit to me (albeit with the initial manual required DNS setup, but that's not a big problem).

**Comment by xenolf** [comment]
I'd like to have all of them implemented in the next 1-2 weeks to conform with the spec on or right after the GA date of let's encrypt.

**Comment by xenolf** [comment]
DNS challenge blocked by letsencrypt/boulder/issues/720

**Comment by jimmycuadra** [comment]
DNS should be unblocked now. letsencrypt/boulder#720 is closed and LE tweeted that DNS is available in production now: https://twitter.com/letsencrypt/status/689919523164721152

**Comment by dpiddy** [comment]
Beware https://github.com/letsencrypt/boulder/issues/1391

**Comment by xenolf** [comment]
DNS-01 is now usable on the library level. However testing indicated that the providers do run into the issue mentioned by @dpiddy. I will add CLI support and a proper release when that issue is fixed on staging and full testing of the functionality was possible.

**Comment by drpebcak** [comment]
@xenolf what is the status of CLI support for dns-01? Looks like the upstream issue was closed, and a fix was merged.

**Comment by xenolf** [comment]
@drpebcak Yes, the staging environment was updated on the 26th with the patch and we ran some tests yesterday. I hope I can get to it tomorrow :)

**Comment by sondr3** [comment]
Awesome, thanks for a great library! :)

**Comment by drpebcak** [comment]
@xenolf awesome, thanks a lot!

**Comment by PunKeel** [comment]
Awesome <3 (should update readme ;D)

**Comment by xenolf** [comment]
CLI support for DNS-01 is now available.

**Comment by xenolf** [comment]
And please keep in mind that the DNS feature along with all providers and the CLI integration is new and I'm pretty sure there are some issues which need to be ironed out :)

**Comment by jimmycuadra** [comment]
Just tried the DNS solver with Route 53 and made it this far:

```
$ lego --email="user@example.com" --domains="example.com" --dns="route53" --exclude="http-01" --exclude="tls-sni-01" run
2016/01/29 23:06:03 [INFO][example.com] acme: Obtaining bundled SAN certificate
2016/01/29 23:06:03 [INFO][example.com] acme: Could not find solver for: http-01
2016/01/29 23:06:03 [INFO][example.com] acme: Could not find solver for: tls-sni-01
2016/01/29 23:06:03 [INFO] acme: Trying to solve DNS-01
2016/01/29 23:06:04 [example.com] Could not obtain certificates
	Error presenting token Request failed, got status code: 400. Response: <?xml version="1.0"?>
<ErrorResponse xmlns="https://route53.amazonaws.com/doc/2013-04-01/"><Error><Type>Sender</Type><Code>InvalidChangeBatch</Code><Message>Invalid Resource Record: FATAL problem: InvalidTXTRDATA encountered at hcpTWGxkPEGUMxJMPsXtsPE0TSflliWqAKH1FJ8crvg</Message></Error><RequestId>REDACTED</RequestId></ErrorResponse>
```

I think the DNS TXT record needs to have literal double quotes around it as part of the value—that may be the issue.

Also, it should be added to the CLI help message that you also need to supply the environment variable AWS_REGION or you will get an error. It was also unclear that I needed to specifically exclude the other solvers to use the DNS solver, because it tries them each in order even if you pass CLI flags for the DNS solver.

**Comment by jimmycuadra** [comment]
It might also be useful to document which AWS calls are used for people that want to create an IAM policy for lego. Here is what worked for me:

``` json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Sid": "Stmt1454129920000",
            "Effect": "Allow",
            "Action": [
                "route53:ChangeResourceRecordSets"
            ],
            "Resource": [
                "arn:aws:route53:::hostedzone/<INSERT_YOUR_HOSTED_ZONE_ID_HERE>"
            ]
        },
        {
            "Sid": "Stmt1454129944000",
            "Effect": "Allow",
            "Action": [
                "route53:ListHostedZones"
            ],
            "Resource": [
                "*"
            ]
        }
    ]
}
```

**Comment by xenolf** [comment]
@jimmycuadra could you try again using the commit 6e423e6c7ac3668725bc0aadf54ebfa24f8064a1 ?
If not, could you open a new issue to track this?

I also added the AWS_REGION env variable to the README.

**Comment by fenderle** [comment]
Any estimation when POP is implemented? Am I right that this challenge will neither need an external accessible server nor DNS Updates?

**Comment by mholt** [comment]
@fenderle I believe that's correct. Honestly it's probably not a priority right now since no ACME CA supports it (yet). But I'm sure once DNS is working better it'd be nice to just "finish" this package and have PoP working even if it's not enabled for now.

**Comment by xenolf** [comment]
Right, I did not invest any time into PoP yet as we can't use it anyways. We will support it when it's available though.

**Comment by xenolf** [comment]
The PoP challenge has been removed from the ACME spec. Closing this issue.
